### ISSUE TITLE
一些问题

### ISSUE THREAD

**Issue by 00ffcc** [YuchuanTian/DiJiang#5]
1. 文章中说  $\Phi_{WDCF}(x)=De^{TCx^T}$ ,但是代码里用的是softmax, (query = nn.functional.softmax(query@self.proj_matrix, dim=-1)), 这样是等价的吗？
2. 公式12中的 $\Phi_{WDCF}$ 和公式10中的 $\phi_{WPFF}$ 有什么关系呢？

谢谢！希望能得到解答

**Comment by HantingChen** [comment]
你好，变换的完整形式为：       
query = nn.functional.softmax(query@self.proj_matrix, dim=-1)
query = query*self.D1[:,:,:query_length,:]

我们在进行指数计算后为了保证数值稳定性，进行了归一化，等价于直接使用softmax，D对应着self.D1，TC对应self.proj_matrix。

公式12中的WDCF是WPFF使用了频域变换的变体，即将self.proj_matrix中使用DCT变换进行计算，可以使得计算更高效。

**Comment by 00ffcc** [comment]
不需要做IDCT吗? 还有进行归一化操作之后还能满足无偏估计的性质吗？谢谢

**Comment by 00ffcc** [comment]
能展开讲讲是怎么得到频域变换的变体的吗？👀

**Comment by HantingChen** [comment]
> 不需要做IDCT吗? 还有进行归一化操作之后还能满足无偏估计的性质吗？谢谢

不需要做IDCT，可以满足无偏估计的性质。

**Comment by HantingChen** [comment]
> 能展开讲讲是怎么得到频域变换的变体的吗？👀

直接将随机映射用频域变换进行替代即可。
